authors/apps/articles/tests/test_rate_article.py

```diff
@@ -0,0 +1,186 @@
+            format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_unathenticated_user_can_see_ratings(self):
+        self.client.logout()
+        response = self.client.get(
+            f'/api/articles/rate/{slug}/',
+            {"article": valid_rating},
+            format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIn(slug, response.data['slug'])
+
+    def test_athenticated_user_can_see_ratings(self):
+        response = self.client.get(
+            f'/api/articles/rate/{slug}/',
+            {"article": valid_rating},
+            format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIn(slug, response.data['slug'])
+
+    def test_user_cant_rate_their_article(self):
+        response = self.client.post(
+            f'/api/articles/rate/{slug}/',
+            {"article": valid_rating},
+            format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+        self.assertIn(
+            'you cant rate your own article', response.data['ratings']
+            )
+
+    def test_auth_user_can_rate_an_article(self):
+        self.client.post('/api/articles/',
+            {"article": valid_article}, format="json")
+        self.client.logout()
+        response = self.client.post(
+            register_url, {"user": valid_user2}, format="json")
+
+        self.token = response.data["data"]["token"]
+
+        self.client.get(f"{activate_url}?token={self.token}", format='json')
+
+        self.client.credentials(HTTP_AUTHORIZATION='Token ' + self.token)
+        self.client.post(
+            login_url, {"user": valid_user2}, format='json')
+
+        response = self.client.post(
+            f'/api/articles/rate/{slug}/',
+            {"article": valid_rating},
+            format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_one_can_only_rate_from_0_to_5(self):
+        self.client.logout()
+        user = User.objects.create_user(**valid_user2)
+        self.client = APIClient()
+        self.client.force_authenticate(user=user)
+        self.client.login(**valid_user2)
+        response = self.client.post(
+            f'/api/articles/rate/{slug}/',
+            {"article": invalid_rating},
+            format="json"
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(
+            'only digits from 0 to 5 are allowed', response.data['ratings']
+            )
+
+    def test_one_can_only_rate_from_0_to_5(self):
+        self.client.logout()
+        user = User.objects.create_user(**valid_user2)
+        self.client = APIClient()
+        self.client.force_authenticate(user=user)
+        self.client.login(**valid_user2)
+        response = self.client.post(
```

Comment:
```
self.client.logout()
user = User.objects.create_user(**valid_user2)
self.client = APIClient()
self.client.force_authenticate(user=user)
self.client.login(**valid_user2)
```
You are using this pattern on almost every your test function hence making alot of code duplication. I suggest you create that user and login them once in your setup function.

Reply:
on it

Reply:
I have fixed this